File: 0731-my-calendar-ii/0731-my-calendar-ii.py
class MyCalendarTwo:

    def __init__(self):
        self.calender=[]
        self.overlap=[]

    def book(self, start: int, end: int) -> bool:
        for s,e in self.overlap:
            if(start<e and end > s):
                return False
        for s,e in self.calender:
            if(start < e and end> s):
                cs=max(start,s)
                ce=min(end,e)
                self.overlap.append((cs,ce))
        self.calender.append((start,end))
        return True


# Your MyCalendarTwo object will be instantiated and called as such:
# obj = MyCalendarTwo()
# param_1 = obj.book(start,end)

File: 0731-my-calendar-ii/0731-my-calendar-ii.java
class MyCalendarTwo {

       private List<int[]> bookings;
        private List<int[]> overlaps;
    
    
    public MyCalendarTwo(){
        overlaps= new ArrayList<>();
        bookings= new ArrayList<>();
    }
    
    public boolean book(int start, int end) {
        for(int[] overlap: overlaps){
            if(start< overlap[1] && end > overlap[0]){
                return false;
            }
        }
        for(int[] booking: bookings){
            if(start< booking[1] && end > booking[0]){
                int os= Math.max(start,booking[0]);
                int oe= Math.min(end,booking[1]);
                overlaps.add(new int[]{os,oe});
            }
        }
        bookings.add(new int[]{start,end});
        return true;
        
    }
}

/**
 * Your MyCalendarTwo object will be instantiated and called as such:
 * MyCalendarTwo obj = new MyCalendarTwo();
 * boolean param_1 = obj.book(start,end);
 */